### ISSUE TITLE
http -> https redirect bug

### ISSUE THREAD

**Issue by tcolgate** [traefik/traefik#8345]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.containo.us) and didn't find any.

### What did you do?

configured traefik on k8s with
```yml
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: redirect-http-https
  namespace: infra
spec:
  redirectScheme:
    permanent: true
    scheme: https
```

expected only http calls to be redirected from http -> https

### What did you see instead?

If a client includes a port number in the http header (not sure if this is spec compliant, but some of our are), traefik sends a redirect for calls which are already using https. e.g.

`curl -v -H "Host: atom-content-staging.qutics.com:443"  https://atom-content-staging.qutics.com/q`

gets sent a  redirect, even though it can in with the correct scheme.

### What version of Traefik are you using?

2.4.8 from docker

### What is your environment & configuration?

```yaml
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: redirect-http-https
  namespace: infra
spec:
  redirectScheme:
    permanent: true
    scheme: https
```

This is added to a regular networking/v1 ingress with an annotation.

**Comment by jbdoumenjou** [comment]
Hello @tcolgate,

Could you please share your complete configuration (with the ingress) and the Traefik log from the beginning ?

**Comment by tcolgate** [comment]
Hi,
Our config is rather complicated (two instances of traefik process different ingresses, middlewares are shared). Most of it is irrelevant to the problem at hand though.
This is the main config file.
```
defaultEntryPoints = []
[entryPoints]
  [entryPoints.http-public]
  address = ":8880"
  [entryPoints.http-public.transport.lifecycle]
  graceTimeOut= "10s"
  requestAcceptGraceTimeOut= "30s"

  [entryPoints.https-public]
  address = ":443"
  [entryPoints.https-public.http.tls]
  [entryPoints.https-public.transport.lifecycle]
  graceTimeOut= "10s"
  requestAcceptGraceTimeOut= "30s"

[tracing.jaeger]
localAgentHostPort = "{{ env "KUBERNETES_HOST_IP" }}:6831"
samplingType = {{ metadata "samplingType" | quote }}
samplingParam = {{ metadata "samplingParam" }}

[web]
address = ":8080"
ReadOnly = true

[web.statistics]
RecentErrors = 10

[respondingTimeouts]
idleTimeout = "10s"

[api]
dashboard = true
insecure = true

[ping]

[metrics]
  [metrics.prometheus]

[accessLog]
format = "json"

[serversTransport]
  rootCAs = [ "/config/ca.crt" ]
  maxIdleConnsPerHost = 20

[providers.kubernetescrd]
allowcrossnamespace = true
ingressclass  = "traefik"
labelselector = "externalAccess in (public,traefikSharedResource)"

[providers.kubernetesingress]
ingressclass  = "traefik"
labelselector = "externalAccess in (public,traefikSharedResource)"

[log]
format = "json"
level = "fatal"

```
middlware is:
```
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  labels:
    externalAccess: traefikSharedResource
  name: redirect-http-https
  namespace: infra
spec:
  redirectScheme:
    permanent: true
    scheme: https
```
(there also a retry middleware, but which is also not relevant)

ingresses are configured using the regular kube ingress plus annotations

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: traefik
    traefik.ingress.kubernetes.io/router.middlewares: infra-retry@kubernetescrd,infra-redirect-http-https@kubernetescrd
  labels:
    externalAccess: public
  name: someapp
  namespace: default
spec:
  rules:
  - host: someapp.example.com
    http:
      paths:
      - backend:
          service:
            name: blocker
            port:
              name: http
        path: /debug
        pathType: Prefix
      - backend:
          service:
            name: someapp
            port:
              name: http
        path: /
        pathType: Prefix

```

I can't run the impacted instances with debug logging due to traffic volumes, but the actual bug is fairly self evident in the code. The link PR include a test which demonstrates the problem.

Some of our clients are setting `Host: appname.example.com:443` in the host header. This gets passed (including the port) into req.Host, The test to determine if the old URL matches the new URL then fails because the new URL evaluates to "https://appname.example.com/", and the where the old name evaluates to "https://appname.example.com:443/".

**Comment by jbdoumenjou** [comment]
Thanks.

Sorry, to be clearer, I should only have asked for the complete configuration for a minimal reproducible case. :smile:

**Comment by tcolgate** [comment]
It'll take me a while to put together a minimal config, but anything that's using the RedirectScheme middleware be impacted.

**Comment by tcolgate** [comment]
We need to be able to enable redirection on a per-ingress basis.

**Comment by jbdoumenjou** [comment]
Just to understand better, why do you need per-ingress redirection instead of by Entrypoint ?

**Comment by tcolgate** [comment]
Because  we have existing customers that hit some of our older services with clients that wont respect redirects. We want stuff to redirect by default, but for some specific cases, we need to disable it and accept the traffic directly.

I've tested the fix that I've PR'd and it seems to work.

**Comment by tcolgate** [comment]
FWIW, the first thing that actually triggered this bug was a grpc client (part of thanos), that failed talking via traefik to another component that sits behind traefik.

**Comment by mpl** [comment]
@tcolgate the PR ( #8348 ) was reworked because we think the root cause of the problem was different from what you seemed to be describing. Could you please test it and let us know if it works for you?

**Comment by tcolgate** [comment]
@mpl the reworked PR looks fine, though It's not immediately obvious that the tests cover the problem cases (unless the contention is that the value of req.Host is irrelevant, but that code does do the host/port split on req.Host, so not setting req.Host somehow in the tests fees incomplete)

**Comment by traefiker** [comment]
Closed by #8348.
